src/lib/int.ts

```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
-    } else {
-      return '-' + this.value.neg().toString();
-    }
+  static get zero() {
+    return new Int64(UInt64.zero, Field.one);
   }
-
-  static get zero(): Int64 {
-    return new Int64(Field.zero);
+  static get one() {
+    return new Int64(UInt64.one, Field.one);
   }
-
-  static fromUnsigned(x: UInt64): Int64 {
-    return new Int64(x.value);
+  static get minusOne() {
+    return new Int64(UInt64.one, Field.one.neg());
   }
 
-  private static shift(): Field {
-    return Field.fromJSON((1n << 64n).toString()) as Field;
+  toField() {
+    return this.magnitude.value.mul(this.sgn);
   }
 
-  uint64Value(): Field {
-    const n = BigInt(this.value.toString());
-    if (n < 1n << 64n) {
-      return this.value;
-    } else {
-      const x = this.value.add(Int64.shift());
-
-      return x;
-    }
+  static fromField(x: Field): Int64 {
+    let getUnchecked = () => Int64.fromFieldUnchecked(x);
+    // constant case - just return unchecked value
+    if (x.isConstant()) return getUnchecked();
+    // variable case - create a new checked witness and prove consistency with original field
+    let xInt = Circuit.witness(Int64, getUnchecked);
+    xInt.toField().assertEquals(x); // sign(x) * |x| === x
+    return xInt;
   }
 
-  static sizeInFields(): number {
-    return 1;
+  neg() {
+    // doesn't need further check if `this` is valid
+    return new Int64(this.magnitude, this.sgn.neg());
   }
 
-  neg(): Int64 {
-    return new Int64(this.value.neg());
+  add(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().add(y_.toField()));
   }
-
-  static add(x: Int64, y: Int64 | UInt32 | UInt64) {
-    return new Int64(x.value.add(y.value));
+  sub(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().sub(y_.toField()));
   }
-
-  static sub(x: Int64, y: Int64 | UInt32 | UInt64) {
-    return new Int64(x.value.sub(y.value));
+  mul(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().mul(y_.toField()));
   }
-
-  add(y: Int64 | UInt32 | UInt64) {
-    return Int64.add(this, y);
+  div(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    let [q] = this.magnitude.divMod(y_.magnitude);
+    let sign = this.sgn.mul(y_.sgn);
+    return new Int64(q, sign);
   }
-
-  sub(y: Int64 | UInt32 | UInt64) {
-    return Int64.add(this, y);
+  mod(y: UInt64 | number | string) {
+    let y_ = new UInt64(argToField('Int64.mod', y));
+    let [, r] = this.magnitude.divMod(y_);
+    r = Circuit.if(this.isPositive(), r, y_.sub(r));
+    return new Int64(r, Field.one);
   }
 
-  repr(): { magnitude: Field; isPos: Sgn } {
-    throw 'repr';
+  equals(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return this.toField().equals(y_.toField());
```

Comment:
would `this.magnitude.equals(y_.magnitude).and(this.sign.equals(y_.sign)` be better? The two equalities could be optimized by the compiler (using the permutation), and thus you'd only have one constraint (the `and`)

Reply:
I didn't do that because I was worried about special-casing for 0, where -0 should equal 0.. The special-casing would also require multiplication constraints I think

Reply:
good point :o